### PR TITLE
Add test in scheduler/util/utils_test.go

### PR DESCRIPTION
**Pull request by 196Ikuchil**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Add some tests in scheduler/util/utils_test.go to improve the test coverage.
Modified test cases for multiple pods.
- `TestClearNominatedNodeName `

Add new test because there was no test.
- `TestIsScalarResourceName `

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: pkg/scheduler/util/utils_test.go [hunk 1]
```diff
@@ -252,3 +260,90 @@ func TestPatchPodStatus(t *testing.T) {
...
+			want:         true,
+		},
+	}
+	{
+		for _, test := range tests {
+			if IsScalarResourceName(test.resourceName) != test.want {
+				t.Errorf("%s: expected: %t, got: %t", test.name, test.want, !test.want)
+			}
+		}
+	}
```
Comment (sanposhiho) [inline pkg/scheduler/util/utils_test.go:348]:
```suggestion
	for _, test := range tests {
		if IsScalarResourceName(test.resourceName) != test.want {
			t.Errorf("%s: expected: %t, got: %t", test.name, test.want, !test.want)
		}
	}
```

File: pkg/scheduler/util/utils_test.go [hunk 2]
```diff
@@ -123,54 +125,61 @@ func TestMoreImportantPod(t *testing.T) {
 }
 
 func TestRemoveNominatedNodeName(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (196Ikuchil) [reply pkg/scheduler/util/utils_test.go]: Thank you for your advice!
I have added new test cases to cover those.

File: pkg/scheduler/util/utils_test.go [hunk 3]
```diff
@@ -122,55 +124,104 @@ func TestMoreImportantPod(t *testing.T) {
...
-				return true, &v1.Pod{}, nil
+				// If the pod name is "err", return an error.
+				if patch.GetName() == "err1" {
+					return true, nil, test.patchError
+				} else {
+					actualPatchData = append(actualPatchData, string(patch.GetPatch()))
+					// For this test, we don't care about the result of the patched pod, just that we got the expected
+					// patch request, so just returning &v1.Pod{} here is OK because scheduler doesn't use the response.
+					return true, &v1.Pod{}, nil
+				}
```
Comment (sanposhiho) [inline pkg/scheduler/util/utils_test.go:206]: nits:

```suggestion
				} 
				actualPatchData = append(actualPatchData, string(patch.GetPatch()))
				// For this test, we don't care about the result of the patched pod, just that we got the expected
				// patch request, so just returning &v1.Pod{} here is OK because scheduler doesn't use the response.
				return true, &v1.Pod{}, nil
				
```

File: pkg/scheduler/util/utils_test.go [hunk 4]
```diff
@@ -122,55 +124,104 @@ func TestMoreImportantPod(t *testing.T) {
...
 	tests := []struct {
-		name                     string
-		currentNominatedNodeName string
-		newNominatedNodeName     string
-		expectedPatchRequests    int
-		expectedPatchData        string
+		name                  string
+		pods                  []*v1.Pod
+		patchError            error
+		expectedPatchErrors   []error
```
Comment (sanposhiho) [inline pkg/scheduler/util/utils_test.go:137]: change type from `[]error` to `utilerrors.Aggregate` so that we can compare the return value easily.

File: pkg/scheduler/util/utils_test.go [hunk 5]
```diff
@@ -122,55 +124,104 @@ func TestMoreImportantPod(t *testing.T) {
...
+				{
+					ObjectMeta: metav1.ObjectMeta{Name: "foo2"},
+					Status:     v1.PodStatus{NominatedNodeName: "node2"},
+				},
+			},
+			expectedPatchRequests: 2,
+			expectedPatchData:     `{"status":{"nominatedNodeName":null}}`,
+		},
+		{
+			name:                  "Should not make patch request if nominated node is already cleared",
```
Comment (sanposhiho) [inline pkg/scheduler/util/utils_test.go:157]: I think you can remove this test case.

File: pkg/scheduler/util/utils_test.go [hunk 6]
```diff
@@ -122,55 +124,104 @@ func TestMoreImportantPod(t *testing.T) {
...
+			pods:                  []*v1.Pod{},
+			expectedPatchRequests: 0,
 		},
 		{
-			name:                     "Should not make patch request if nominated node is already cleared",
-			currentNominatedNodeName: "",
-			expectedPatchRequests:    0,
+			name: "Should not be patched if NominatedNodeName is empty",
```
Comment (sanposhiho) [inline pkg/scheduler/util/utils_test.go:162]: Rename it to "Should make patch request only for pods that have NominatedNodeName" and add one pod which have NominatedNodeName to `pods` field.

File: pkg/scheduler/util/utils_test.go [hunk 7]
```diff
@@ -122,55 +124,104 @@ func TestMoreImportantPod(t *testing.T) {
...
+			pods: []*v1.Pod{
+				{
+					ObjectMeta: metav1.ObjectMeta{Name: "foo1"},
+					Status:     v1.PodStatus{NominatedNodeName: "node1"},
+				},
+				{
+					// In this test, the pod named "err1" returns an error in the patch.
+					ObjectMeta: metav1.ObjectMeta{Name: "err1"},
+					Status:     v1.PodStatus{NominatedNodeName: "node1"},
+				},
```
Comment (sanposhiho) [inline pkg/scheduler/util/utils_test.go:182]: Reverse the order so that this test case means "make patch requests for all pods even if a patch for one pods fails"

Comment (sanposhiho) [reply pkg/scheduler/util/utils_test.go:182]: And could you add the comment to describe this behavior to `ClearNominatedNodeName`'s doc?

File: pkg/scheduler/util/utils_test.go [hunk 8]
```diff
@@ -122,55 +124,104 @@ func TestMoreImportantPod(t *testing.T) {
...
+		{
+			name: "Should return error if patch fails",
+			pods: []*v1.Pod{
+				{
+					ObjectMeta: metav1.ObjectMeta{Name: "foo1"},
+					Status:     v1.PodStatus{NominatedNodeName: "node1"},
+				},
+				{
+					// In this test, the pod named "err1" returns an error in the patch.
+					ObjectMeta: metav1.ObjectMeta{Name: "err1"},
```
Comment (sanposhiho) [inline pkg/scheduler/util/utils_test.go:180]: Could you fix the `cs.AddReactor` as well

```suggestion
					// In this test, the pod named "err" returns an error in the patch.
					ObjectMeta: metav1.ObjectMeta{Name: "err"},
```

File: pkg/scheduler/util/utils_test.go [hunk 9]
```diff
@@ -122,55 +124,104 @@ func TestMoreImportantPod(t *testing.T) {
...
-		currentNominatedNodeName string
-		newNominatedNodeName     string
-		expectedPatchRequests    int
-		expectedPatchData        string
+		name                  string
+		pods                  []*v1.Pod
+		patchError            error
+		expectedPatchErrors   []error
+		expectedPatchRequests int
+		expectedPatchData     string
```
Comment (sanposhiho) [inline pkg/scheduler/util/utils_test.go:139]: change type from string to []string so that we can compare the patch data easily.

File: pkg/scheduler/util/utils_test.go [hunk 10]
```diff
@@ -122,55 +124,104 @@ func TestMoreImportantPod(t *testing.T) {
 	}
 }
 
-func TestRemoveNominatedNodeName(t *testing.T) {
+func TestClearNominatedNodeName(t *testing.T) {
+	t.Parallel()
```
Comment (sanposhiho) [inline pkg/scheduler/util/utils_test.go:128]: not need to use `t.Parallel()` in k/k.
Ref: https://github.com/kubernetes/kubernetes/pull/107674#discussion_r817125869

Comment (196Ikuchil) [reply pkg/scheduler/util/utils_test.go]: I didn't know that, thank you!
And I have fixed those.

File: pkg/scheduler/util/utils_test.go [hunk 11]
```diff
@@ -122,54 +126,96 @@ func TestMoreImportantPod(t *testing.T) {
...
 
-			pod := &v1.Pod{
-				ObjectMeta: metav1.ObjectMeta{Name: "foo"},
-				Status:     v1.PodStatus{NominatedNodeName: test.currentNominatedNodeName},
-			}
+			})
 
-			if err := ClearNominatedNodeName(cs, pod); err != nil {
-				t.Fatalf("Error calling removeNominatedNodeName: %v", err)
+			if err := ClearNominatedNodeName(cs, test.pods...); err != nil && err.Is(test.expectedPatchError) {
```
Comment (sanposhiho) [inline pkg/scheduler/util/utils_test.go:210]: Compare errors by `!reflect.DeepEqual(err.Errors(), test.expectedPatchError.Errors())`. (Your approach may miss some wrong test cases.)

File: pkg/scheduler/util/utils_test.go [hunk 12]
```diff
@@ -122,54 +126,96 @@ func TestMoreImportantPod(t *testing.T) {
 	}
 }
 
-func TestRemoveNominatedNodeName(t *testing.T) {
+func TestClearNominatedNodeName(t *testing.T) {
+	statusErr := &apierrors.StatusError{
+		ErrStatus: metav1.Status{Reason: metav1.StatusReasonUnknown},
```
Comment (sanposhiho) [inline pkg/scheduler/util/utils_test.go:131]: nits: `Message` field is used when this status error is formatted and logged.

```suggestion
		ErrStatus: metav1.Status{Reason: metav1.StatusReasonUnknown, Message: "status error"},
```

File: pkg/scheduler/util/utils_test.go (hunk 11 above)
Comment (196Ikuchil) [reply pkg/scheduler/util/utils_test.go]: > avoid using reflect.DeepEqual with errors deepequalerrors

yes, I tried to do it first. But I got the above warning in my code editor. No problem? (The test succeeds.)

Comment (sanposhiho) [reply pkg/scheduler/util/utils_test.go]: What kind of error? I don't see a problem with this approach of comparison.

Comment (sanposhiho) [reply pkg/scheduler/util/utils_test.go]: I don't see a problem with this method of comparison.

Comment (196Ikuchil) [reply pkg/scheduler/util/utils_test.go]: Not error but warning. And warning text is just this.
> avoid using reflect.DeepEqual with errors 

It is displayed only on the code editor (I use vscode) not on the terminal. So, I don't think it's a problem.

**Review by sanposhiho — COMMENTED** [review]
/lgtm

**Comment by 196Ikuchil** [comment]
Thank you!
/sig testing

**Comment by sanposhiho** [comment]
You don't need to add `sig` by yourself in most cases.
/remove-sig testing

**Comment by 196Ikuchil** [comment]
@damemi 
If you have a free moment, could you review this, please?

File: pkg/scheduler/util/utils_test.go [hunk 13]
```diff
@@ -122,54 +126,96 @@ func TestMoreImportantPod(t *testing.T) {
...
+			if err := ClearNominatedNodeName(cs, test.pods...); err != nil && !reflect.DeepEqual(err.Errors(), test.expectedPatchError.Errors()) {
+				t.Fatalf("ClearNominatedNodeName's error mismatch: Actual was %v, but expected %v", err.Errors(), test.expectedPatchError.Errors())
 			}
 
 			if actualPatchRequests != test.expectedPatchRequests {
 				t.Fatalf("Actual patch requests (%d) dos not equal expected patch requests (%d)", actualPatchRequests, test.expectedPatchRequests)
 			}
 
-			if test.expectedPatchRequests > 0 && actualPatchData != test.expectedPatchData {
+			if test.expectedPatchRequests > 0 && !reflect.DeepEqual(actualPatchData, test.expectedPatchData) {
```
Comment (denkensk) [inline pkg/scheduler/util/utils_test.go:218]: change to `cmp.Diff`

File: pkg/scheduler/util/utils_test.go [hunk 14]
```diff
@@ -252,3 +297,47 @@ func TestPatchPodStatus(t *testing.T) {
...
+			resourceName: "requests.foo",
+			want:         false,
+		},
+		{
+			name:         "Should be true with hugePage resource name",
+			resourceName: "hugepages-2Mi",
+			want:         true,
+		},
+		{
+			name:         "Should be false with not HugePage resource name",
```
Comment (denkensk) [inline pkg/scheduler/util/utils_test.go:323]: Feel the name and case seems not related？

HugePage resource name is one of the filter. right?

File: pkg/scheduler/util/utils_test.go [hunk 15]
```diff
@@ -122,54 +126,96 @@ func TestMoreImportantPod(t *testing.T) {
...
 
-			pod := &v1.Pod{
-				ObjectMeta: metav1.ObjectMeta{Name: "foo"},
-				Status:     v1.PodStatus{NominatedNodeName: test.currentNominatedNodeName},
-			}
+			})
 
-			if err := ClearNominatedNodeName(cs, pod); err != nil {
-				t.Fatalf("Error calling removeNominatedNodeName: %v", err)
+			if err := ClearNominatedNodeName(cs, test.pods...); err != nil && !reflect.DeepEqual(err.Errors(), test.expectedPatchError.Errors()) {
```
Comment (denkensk) [inline pkg/scheduler/util/utils_test.go:210]: change to `cmp.Diff`

File: pkg/scheduler/util/utils_test.go (hunk 14 above)
Comment (196Ikuchil) [reply pkg/scheduler/util/utils_test.go]: Thank you for your comments.
I renamed it and added some new test cases for the non-scalar-resources name.

**Comment by 196Ikuchil** [comment]
/assign @ahg-g 
Hi, could you review this, please?

**Comment by kerthcet** [comment]
Can you help to rebase the codes. @196Ikuchil

**Comment by sanposhiho** [comment]
@196Ikuchil Could you rebase the branch?

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR
